Refactor for loop syntax

bricev opened the pull request:
Hi guys!

First, congrats for the amazing job done with the v3.0 refactoring! It's really wonderful :)

I'm currently integrating v3 in an application that used to use v2. A weird bug appeared due to a lot of events binded to our <video> element. 

I noticed that `t.mediaElementdispatchEvent(event)` (https://github.com/johndyer/mediaelement/blob/master/src/js/core/mediaelement.js#L345) is used multiple times and at some point one of the callbacks is undefined. 
Looking at the console for debug, it seems that the variable `il = callbacks.length` and the actual `callbacks[i]` array are not corresponding (maybe due to the fact JavaScript is an asynchronous language). 
Sorry I don't know how to explain myself better (I'm not an expert), but the sure thing is that the pull request has fixed this issue on my side. 

I hope this helped, 
Cheers,

ron666 commented:
Thanks for working on this PR. I'd really appreciate it if you share a URL or the logs you got, because in my experience that has nothing to do with it. The asynchronous side is only when AJAX are being applied, and unless any event is removing another one, this shouldn't make any difference. I just want to make sure that's the problem and we are not just patching the code. Let me know your thoughts, and once it's completely understood the nature of this, I'll merge this fix.

bricev commented:
Sure!
Sure :)

To reproduced the error I've simply added logs to the compiled `mediaelement.js` lib.

Here is the log I've added to mediaelement.js:
https://gist.github.com/bricev/ab0e74dd0745245f67088e69099582e0/05ecacacafa28ec3ee995ed45d76055334d93991

This is what I've got in my client console:
<img width="656" alt="without fix" src="https://cloud.githubusercontent.com/assets/4078270/23831385/3dc25e78-0720-11e7-9787-c88385e5c702.png">

Then I added my fix:
https://gist.github.com/bricev/ab0e74dd0745245f67088e69099582e0/revisions#diff-47f8f92358f2bf2449d26d24a499ad0f

And here is the console:
<img width="651" alt="with fix" src="https://cloud.githubusercontent.com/assets/4078270/23831405/a488f28e-0720-11e7-8df7-be6d0096349a.png">

I really do think `t.mediaElement.dispatchEvent` is called a lot because of the fact we added a lot  of events to our <video> element (at least more than average). It seems that at some point `var il` that save the array length in memory is not corresponding to the actual length of the `callbacks`  array `for` loop is executing. 

Strangely, in my fix (second screenshot) you can see that `var il` is equal to 9. 
As this var is not set anymore in the function, I have no idea where it comes from :s
But at this point I have no more error!

ron666 commented:
Ok here's what I'm gonna do (just because if the approach I took with those variables is causing this issue, I need to replace this everywhere): I'm gonna submit in a minute a change in terms of how I dealt with `i` and `il` and if that works that means I need to avoid doing this. But if not, I'll merge yours.

ron666 commented:
OK download the master branch and test with the approach I took. Only reason I'm asking to do this is because as a best practice, you always need to set the total of what you are looping before you start looping. Let me know ASAP, please.

bricev commented:
Ok I understand what you are doing.
I'm (very) sorry to disappoint but the bug is still there :/

I doubled checked and re-entered my logs in the code coming from your last push.
Result is the same.

I keep thinking that `t.mediaElement.dispatchEvent` is called multiples time in a same fraction of seconds (which I do see from my client console). Therefore, each call is executed asynchronously by JavaScript. 
Now I don't get how vars such as `_i3` and `_total3` are mixed at some point but my logs are clear : the loop pointer responds to a faulty condition where the item count does not correspond to the parsed array.

At least by checking to total items right from the array/object itself each time JS iterates from the loop, you are sure you work on the good instance of the array/object. But again, I'm no JS guru and you surely have a better expertise in that field ;)

ron666 commented:
I see, well I'll merge this PR. Thanks for the time of explaining all of this to me, since it's an odd one.

ron666 commented:
PR merged. Just give it a try one more time please and let me know if everything is working at it should. Thanks.

bricev commented:
I've updated my app and the bug is gone.
Thank you so much!!

ron666 commented:
Thank you for testing this
